runtime/gc_cheney.c

```diff
@@ -0,0 +1,69 @@
+/**
+ * Here we are implementing the Cheney's algorithm as explained in
+ * https://en.wikipedia.org/wiki/Cheney%27s_algorithm.
+ *
+ */
+
+#include "balrt.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <sys/mman.h>
+
+static uint64_t DEFAULT_HEAP_HALF_SIZE = 1073741824; // 1GB
+static uint64_t HEAP_HEADER_SIZE = 8; // in bytes
+
+uint8_t *from_space_ptr;
+uint8_t *to_space_ptr;
+uint8_t *alloc_ptr;
+uint8_t *scan_ptr;
+uint8_t *heap_limit_ptr;
+
+extern void get_roots();
+
+void _bal_init_heap() {
+    int page_size = getpagesize();
+    const char *heap_env = getenv("BAL_HEAP");
+    uint64_t heap_half_size = (heap_env != NULL) ? atoi(heap_env) : DEFAULT_HEAP_HALF_SIZE;
+    // Make sure heap size is a multiple of page size
+    heap_half_size = (heap_half_size / page_size) * page_size;
+
+    // Removing MAP_ANONYMOUS fails the mmap(), check this
+    from_space_ptr = mmap(NULL, heap_half_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+    if (from_space_ptr == MAP_FAILED) {
+        printf("Heap initialization Failed\n");
+        abort();
+    }
+    alloc_ptr = from_space_ptr;
+    heap_limit_ptr = from_space_ptr + heap_half_size;
+
+    to_space_ptr = mmap(NULL, heap_half_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+    if (to_space_ptr == MAP_FAILED) {
+        printf("Heap initialization Failed\n");
+        abort();
+    }
+}
+
+void collect() {
+    printf("%s\n", "Collect");
+    get_roots();
+    // get_roots() -> calls function in gc-roots, should pass rsp as parameter
+    // swap from_space <-> to_space
+    // should update heap_limit_ptr
+}
+
+UntypedPtr _bal_alloc(uint64_t nBytes) {
+    nBytes = nBytes + HEAP_HEADER_SIZE;
```

Comment:
You need to round up nBytes to a multiple of 8.

Reply:
Fixed in https://github.com/ballerina-platform/nballerina/commit/8964a5ea73124ee8d2132121b66b78e9f0be39a4